### ISSUE TITLE
Move the gem provider logic from puppet-hiera to extlib

### ISSUE THREAD

**Issue by rski** [voxpupuli/puppet-extlib#63]
I'm working on https://github.com/voxpupuli/puppet-network/pull/168, which requires using the gem provider. As @bastelfreak pointed out, there are more than one gem providers and the correct one needs to be selected. This is already implemented in puppet-hiera:
https://github.com/voxpupuli/puppet-hiera/blob/master/manifests/params.pp
I wouldn't like to just copy this into puppet-network, plus it might be needed again in the future in another module.
So, would it be possible to move it from puppet-hiera to extlib? Is it the right place to move it to or is there a better one instead?

**Comment by bastelfreak** [comment]
we've this logic block copy'n pasted into several modules. we should place it here and make a fact out of it?

**Comment by ffrank** [comment]
Can this become a custom function rather than a fact? I'd like to avoid more Ruby code on each and every agent in favor of centralized logic on the master.

**Comment by rnelson0** [comment]
I don't think so, because it needs to be evaluated on the agent, not the master.

**Comment by ffrank** [comment]
As far as I can tell, all that's needed from the agent are facts like `is_pe`, `pe_version`, or `puppetversion`. The current implementation does the computation in a `params` class, which cannot be readily shared between modules. *But* it happens master-side (in the manifest compiler), unless I'm mistaken. So it should be possible to boil down the respective `params` classes to invoke a (common) function that returns the appropriate `gem` provider based on those fact values, or perhaps a hash of all the values that can be helpful for the different modules. (Each module will still need to extract the hash values into discrete variables.)

**Comment by b4ldr** [comment]
This seems old and as its talking about the `puppet-hiera` im guessing now obsolete?
